### PR TITLE
Implement "canonical template sources"

### PR DESCRIPTION
**Pull request by drevell**
A canonical template source is one that isn't tied to the machine where it's running. For example, a template source like
`github/abcxyz/abc/t/rest_server` is canonical because anyone can reference it. A template source like "/home/dave/my_unpacked_tarball" is not canonical because that directory only exists on my machine. As a complicated special case, a path like `/home/dave/git/myrepo` may or may not be canonical, depending on whether the destination is in the same git repo as the source. See the internal design doc (ping me for a link).

Also, we add a new common function `IsStatNotExist()` to factor out the common pattern of checking whether we Stat()ed a path that doesn't exist.

### CODE REVIEW

File: templates/commands/render/render.go [hunk 1]
```diff
@@ -638,28 +638,31 @@ func loadSpecFile(ctx context.Context, fs common.FS, templateDir string) (*spec.
 // Downloads the template and returns the name of the temp directory where it
```
Comment (capri-xiyue) [inline templates/commands/render/render.go:638]: Nits: update the comment?

**Review by capri-xiyue — APPROVED** [review]

File: templates/commands/render/render.go (hunk 1 above)
Comment (drevell) [reply templates/commands/render/render.go]: Fixed.

**Comment by sethvargo** [comment]
@drevell I'm probably missing something obvious, but why isn't a canonical source a function of the downloader?

**Comment by drevell** [comment]
> @drevell I'm probably missing something obvious, but why isn't a canonical source a function of the downloader?

Perhaps that would make more sense. As it is, it seems like it belongs equally well in ParsedSource or Downloader. But I suppose if it were in the Downloader, we could do smarter things like notice redirects. Did you have some rationale in mind?

**Comment by sethvargo** [comment]
> Perhaps that would make more sense. As it is, it seems like it belongs equally well in ParsedSource or Downloader. But I suppose if it were in the Downloader, we could do smarter things like notice redirects. Did you have some rationale in mind?

Redirects were the exact thing I had in mind. I also thought, since the download is already an interface, we could easily check if it's "canonical" by having a private `isCanonical` function on a different interface.

**Comment by drevell** [comment]
> Redirects were the exact thing I had in mind. I also thought, since the download is already an interface, we could easily check if it's "canonical" by having a private `isCanonical` function on a different interface.

Sounds good, done.

**Review by capri-xiyue — APPROVED** [review]

**Pull request merged**